File: sem1/hw8/8.1/list.cpp
#include <iostream>
#include "list.h"
using namespace std;

struct ListElement
{
    int number;
    int value;
    ListElement *next;
};

struct List
{
    ListElement *first;
};

List *createList()
{
    return new List {nullptr};
}

void addElement(List *list, int number, int value)
{
    ListElement *newElement = new ListElement {number, value, list->first};
    list->first = newElement;
}

int findElement(List *list, int number)
{
    ListElement *current = list->first;
    while (current)
    {
        if (current->number == number)
        {
            return current->value;
        }
        current = current->next;
    }
    return -1;
}

void deleteList(List *list)
{
    while (list->first)
    {
        ListElement *current = list->first;
        list->first = current->next;
        delete current;
    }
    delete list;
}

File: sem2/hw1/1.2/List.java
package com.AntonChernikov.g144;

/**
 * Single linked list
 * */
public class List {
    private class ListElement {
        private int value;
        private ListElement next;

        private ListElement(int value, ListElement next) {
            this.value = value;
            this.next = next;
        }
    }

    private ListElement first = null;
    private int size = 0;

    public void addElement(int value) {
        ListElement newElement = new ListElement(value, first);
        first = newElement;
        size++;
    }

    public void deleteElement(int value) {
        if (first.value == value) {
            size--;
            first = first.next;
            return;
        }
        ListElement current = first;
        while (current.next != null) {
            if (current.next.value == value) {
                size--;
                current.next = current.next.next;
                return;
            }
            current = current.next;
        }
    }

    public int size() {
        return size;
    }

    public void print() {
        if (first == null) {
            return;
        }
        ListElement current = first;
        while (current != null) {
            System.out.print(current.value + " ");
            current = current.next;
        }
        System.out.println();
    }
}
